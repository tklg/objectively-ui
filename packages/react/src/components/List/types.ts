import { ForwardRefExoticComponent } from 'react'
import { PropsWithChildrenAndClassName } from 'src/types/PropsWithChildrenAndClassName'

export interface ListProps extends PropsWithChildrenAndClassName {
  ref?: React.RefObject<HTMLUListElement>;
  compact?: boolean;
  divided?: boolean;
  noPadding?: boolean;
}

export interface ListContextProps {
  compact: boolean;
  divided: boolean;
}

export type ListComponent<P = ListProps> = ForwardRefExoticComponent<P>
