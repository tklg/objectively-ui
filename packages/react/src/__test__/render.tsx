import { CssVarsProvider } from 'src/providers'
import { render, RenderOptions } from '@testing-library/react'
import { FC, PropsWithChildren, ReactElement } from 'react'

const AllTheProviders: FC<PropsWithChildren> = ({children}) => {
  return <CssVarsProvider>{children}</CssVarsProvider>
}

const customRender = (
  ui: ReactElement,
  options?: Omit<RenderOptions, 'wrapper'>,
) => render(ui, { wrapper: AllTheProviders, ...options })

export * from '@testing-library/react'
export { customRender as render }
