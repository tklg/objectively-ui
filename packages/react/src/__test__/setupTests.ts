import '@testing-library/jest-dom/extend-expect'
import matchMediaMock from 'src/__test__/matchMedia.mock'
import { createSerializer } from '@emotion/jest'

matchMediaMock()

// configures @emotion/jest to not insert styles
expect.addSnapshotSerializer(createSerializer({
  includeStyles: false,
}))
