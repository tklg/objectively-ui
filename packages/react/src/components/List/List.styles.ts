import { css } from '@emotion/react'
import { ColorTheme } from 'src/types/ColorTheme'
import { PROJECT_SHORTNAME } from 'src/utils/constants'

export const listStyles = (theme: ColorTheme) => css({
  listStyle: 'none',
  margin: 0,
  padding: `${theme.spacing.sm} 0`,

  [`&.${PROJECT_SHORTNAME}-List-noPadding`]: {
    padding: 0,
  },
})
