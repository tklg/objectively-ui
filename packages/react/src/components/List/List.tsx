import { createContext, forwardRef, useMemo } from 'react'
import { listStyles } from 'src/components/List/List.styles'
import { ListComponent, ListContextProps, ListProps } from 'src/components/List/types'
import { useTheme } from 'src/hooks'
import { buildClassName } from 'src/utils/buildClassName'

const ELEMENT_NAME = 'List'

export const ListContext = createContext<ListContextProps>({
  compact: false,
  divided: false,
})

export const List = forwardRef<HTMLUListElement, ListProps>(({
  children,
  className: _className,
  compact,
  divided,
  noPadding,
}, ref) => {
  const theme = useTheme()
  const className = buildClassName(ELEMENT_NAME, {
    compact,
    divided,
    noPadding,
  }, _className)

  const contextValue = useMemo(() => ({
    compact: compact ?? false,
    divided: divided ?? false,
  }), [compact])

  return (
    <ListContext.Provider value={contextValue}>
      <ul
        ref={ref}
        className={className}
        css={listStyles(theme)}
      >
        {children}
      </ul>
    </ListContext.Provider>
  )
}) as ListComponent

if (process.env.NODE_ENV !== 'production') {
  List.displayName = ELEMENT_NAME
}
