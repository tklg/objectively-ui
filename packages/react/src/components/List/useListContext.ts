import { useContext } from 'react'
import { ListContext } from 'src/components/List/List'

export const useListContext = () => useContext(ListContext)
