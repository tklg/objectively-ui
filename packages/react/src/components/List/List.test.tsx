import { render, screen } from 'src/__test__/render'
import { List } from './List'

describe('List', () => {
  it('renders an empty list', async () => {
    render(
      <List />,
    )

    expect(screen.getByRole('list')).toMatchSnapshot()
  })
})
