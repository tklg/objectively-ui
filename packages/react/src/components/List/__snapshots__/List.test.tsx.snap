// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List renders an empty list 1`] = `
<ul
  class="ObjUI-List emotion-0"
/>
`;
